onboarding.payu/src/main/java/com/onboarding/payu/service/impl/PurchaseOrderImpl.java

```diff
@@ -0,0 +1,308 @@
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+/**
+ * Implementation of {@link IPurchaseOrder} interface.
+ *
+ * @author <a href='julian.ramirez@payu.com'>Julian Alberto Ramirez Osorio</a>
+ * @version 1.0.0
+ * @since 1.0.0
+ */
+@Slf4j
+@Service
+public class PurchaseOrderImpl implements IPurchaseOrder {
+
+	private IPurchaseOrderRepository iPurchaseOrderRepository;
+
+	private IProductService iProductService;
+
+	private IOrderProductService iOrderProductService;
+
+	private ICustomerService iCustomerService;
+
+	private PurchaseOrderMapper purchaseOrderMapper;
+
+	@Autowired
+	public PurchaseOrderImpl(final IPurchaseOrderRepository iPurchaseOrderRepository,
+							 final IProductService iProductService,
+							 final IOrderProductService iOrderProductService,
+							 final ICustomerService iCustomerService,
+							 final PurchaseOrderMapper purchaseOrderMapper) {
+
+		this.iPurchaseOrderRepository = iPurchaseOrderRepository;
+		this.iProductService = iProductService;
+		this.iOrderProductService = iOrderProductService;
+		this.iCustomerService = iCustomerService;
+		this.purchaseOrderMapper = purchaseOrderMapper;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Transactional
+	@Override public PurchaseOrderResponse addPurchaseOrder(final PurchaseOrderRequest purchaseOrderRequest) {
+
+		log.debug("addPurchaseOrder(PurchaseOrderDTO) : ", purchaseOrderRequest.toString());
+		final List<Product> productList = getProductsByIds(purchaseOrderRequest.getProductList());
+		final Customer customer = iCustomerService.findById(purchaseOrderRequest.getCustomer().getIdCustomer());
+
+		isValidOrder(productList, purchaseOrderRequest.getProductList());
+		final PurchaseOrder purchaseOrder = iPurchaseOrderRepository.save(purchaseOrderMapper.toPurchaseOrder(customer, productList,
+																											  purchaseOrderRequest));
+		final List<OrderProduct> orderProductList = getOrderProducts(purchaseOrderRequest.getProductList(), productList, purchaseOrder);
+		iOrderProductService.saveAll(orderProductList);
+		orderProductList.stream().forEach(orderProduct -> iProductService
+				.updateStockById(subtract.applyAsInt(orderProduct.getProduct().getStock(), orderProduct.getQuantity()),
+								 orderProduct.getProduct().getIdProduct()));
+
+		return purchaseOrderMapper.toPurchaseOrderResponse(purchaseOrder);
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override public PurchaseOrder findByIdPurchaseOrder(final Integer idPurchaseOrder) {
+
+		return iPurchaseOrderRepository.findById(idPurchaseOrder).orElseThrow(
+				() -> new BusinessAppException(ExceptionCodes.PURCHASE_ORDER_INVALID));
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override public void updateStatusById(final String status, final Integer id) {
+
+		iPurchaseOrderRepository.updateStatusById(status, id);
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Transactional
```

Comment:
This @Transactional anotation is really needeed?

Reply:
It is required because I have a method to update the purchase order by id, without the Transaction annotation, the code failed at runtime.